docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+
+For Elastic Training we will always pass the following parameters:
+
+```bash
+--rdzv-backend=c10d
+
+--rdzv-id will be set automatically.
+
+--rdzv-endpoint will always point to the node-0 Pod.
+```
+
+Since the [etcd and etcd-v2 are legacy rendezvous](https://pytorch.org/docs/stable/elastic/run.html#note-on-rendezvous-backend),
+we won't support them in `TorchSpec`. We can introduce them in the future if users will require them.
+
+```golang
+// TorchSpec represents the configuration for PyTorch.
+type TorchSpec struct {
+
+    // Number of Procs per Node.
+    NumProcPerNode int `json:"numProcPerNode,omitempty"`
+
+    // Used for single-node multi-worker training
+    Standalone bool `json:"standalone,omitempty"`
+
+    // Torch Elastic Policy.
+    ElasticPolicy *TorchElasticPolicy `json:"elasticPolicy,omitempty"`
+}
+
+// If the Elastic Policy is set, the numNodes parameter is ignored.
+// --nnodes=minNodes:maxNodes
+type TorchElasticPolicy struct {
+
+    // The limits to restart TrainJob.
+    // Insert it to the JobSet.spec.failurePolicy.maxRestarts
+    MaxRestarts *in32 `json:"maxRestarts,omitempty"`
+
+    // Min number of nodes for HPA and torchrun.
+    MinNodes *in32 `json:"minNodes,omitempty"`
+
+    // Max number of nodes for HPA and torchrun.
+    MaxNodes *in32 `json:"maxNodes,omitempty"`
+
+    // Metrics for scale up and down replicas.
+    Metrics []autoscalingv2.MetricSpec `json:"metrics,omitempty"`
+}
+```
+
+### The MPI Spec API
+
+The `MPISpec` API represents the configuration for training using MPI orchestration.
+E.g. creation of host-files and SSH keys. Using MPI might be more efficient for training on HPC
+clusters or for some ML frameworks (e.g. [MLX distributed with MPI](https://ml-explore.github.io/mlx/build/html/usage/distributed.html)).
+
+We will fully migrate to the MPI Operator V2 functionality as part of this KEP.
+Check [the proposal for the MPI V2 APIs.](https://github.com/kubeflow/mpi-operator/blob/master/proposals/scalable-robust-operator.md)
+
+```golang
+type MPISpec struct {
+    // Number of Procs per Node.
+    NumProcPerNode int `json:"numProcPerNode,omitempty"`
+
+    // MPI Implementation to create appropriate host-files.
+    // Can be one of OpenMPI, Intel, or MPICH.
+    MPIImplementation MPIImplementation `json:"mpiImplementation,omitempty"`
+
+    // Directory where SSH keys are mounted.
+    SSHAuthMountPath string `json:"SSHAuthMountPath,omitempty"`
+}
+
+type MPIImplementation string
+
+const (
+    MPIImplementationOpenMPI MPIImplementation = "OpenMPI"
+    MPIImplementationIntel   MPIImplementation = "Intel"
+    MPIImplementationMPICH   MPIImplementation = "MPICH"
+)
+```
+
+### Supported Runtimes by Community
+
+Kubeflow community are planning to support the following runtimes.
+
+#### PyTorch Distributed Runtime
+
+Initially, we will maintain only multi-node multi-worker runtime and PyTorch Elastic.
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: ClusterTrainingRuntime
```

Comment:
Yeah thinking this through more - let's talk about why different abstractions are needed for cluster vs. not. I think this design would be much simpler (easier for the user to use) if there were just one.

Reply:
Agreed with the idea that having two layers is complex.

Reading through the doc, I think ClusterTrainingRuntime is meant to be something set by Kubeflow community or designed by platform engineers in an org incorporating all the best practices. This is cluster scoped. This may not be runnable as is and needs to be further tuned or scoped down for team specific settings like model provider or custom init containers. 

ML teams may just have access to namespace-scoped resources so they end up copying the blueprint, fill in more defaults and make them available with other MLEs in their team to try with different trainerConfigs. Since these blueprints are specific to a particular team of engineers, it may not fit the cluster scope. This part needs to be optional however ie. TrainerConfigs can reference a cluster-scoped ClusterTrainingRuntime.

The idea does have merits (similar to ClusterRole and Role concepts) but is certainly increases complexity.

Reply:
@tenzen-y Can share concrete on-prem use-case why do we need `TrainingRuntime` and `ClusterTrainingRuntime`.

TLDR: If you want to share blueprint between all Kubernetes namespaces, you might want to use `ClusterTrainingRuntime`.

I think, that gives flexibility for Platform Engineers to configure blueprints in the Kubernetes cluster.
For example, if you want to create unify MPI runtime per cluster, you should use `ClusterTrainingRuntime`.
On the other side, if you have different namespaces in your cluster (e.g. `ad-hoc` and `batch`), you might consider to use different runtimes for every namespace. For example, using different accelerators or node selectors.

I understand your confusion in this comment: https://github.com/kubeflow/training-operator/pull/2171#discussion_r1680385287, but let's continue discussion in this 🧵 to see the value of both runtimes.


A few reference:
- KServe `ServingRuntime` and `ClusterServingRuntime`: https://kserve.github.io/website/latest/modelserving/servingruntimes/
- Cert Manager `Issuer` and `ClusterIssuer`: https://cert-manager.io/docs/concepts/issuer/#namespaces

@ahg-g would love to get your feedback as well.

Reply:
Some use cases here: 
- `ClusterTrainingRuntime` are defined by a cluster admin or platform developers and then researchers specify the clusterTrainingRuntime name in the TrainJob API.
- `TrainingRuntime` are defined by researchers within the their namespaces (tenants).

If we do not provide the `ClusterTrainingRuntime` API, admins need to create the TrainingRuntime API in all user namespaces. It is terrible since the multitenant cluster has hundred and thousand namespases.